File: bindings/python/examples/basic_messaging.py
"""Basic messaging: create identity, create context, send and receive messages.

Post-Phase-4 (#1549, ADR-048): every SDK call routes through an explicit
:class:`scp_sdk.SCP` instance. Construct one at process start, call
:meth:`SCP.identity_create` / :meth:`SCP.context_create` / etc., and let
the ``with`` block drain in-flight work via ``scp.shutdown()`` on exit.
"""

import asyncio

from scp_sdk import SCP
from scp_sdk.types import Capability, CustodyType, MemoryScope


async def main() -> None:
    with SCP(storage={"type": "in_memory"}) as scp:
        # Create two identities in the encrypted key file SCP implements.
        alice = await scp.identity_create(CustodyType.ENCRYPTED_FILE)
        bob = await scp.identity_create(CustodyType.ENCRYPTED_FILE)
        print(f"Alice DID: {alice.did}")
        print(f"Bob DID: {bob.did}")

        # Alice creates a context.
        ctx = await scp.context_create(
            alice.did,
            {
                "ceiling": [
                    Capability.MESSAGES_READ.value,
                    Capability.MESSAGES_WRITE.value,
                    Capability.MEMBER_INVITE.value,
                ],
                "memory_scope": MemoryScope.EPHEMERAL.value,
                "governance": "single_admin",
                "ttl": 3600.0,
            },
        )
        print(f"Context ID: {ctx.context_id}")

        # Bob joins the context.
        await scp.context_join(ctx._raw_handle, bob.did)
        print(f"Bob joined: {bob.did}")

        # Alice sends a message.
        await scp.context_send(ctx._raw_handle, alice.did, b"Hello Bob, this is Alice")

        # Bob receives it.
        receiver = await scp.context_receive(ctx._raw_handle)
        async for msg in receiver:
            print(f"Bob received from {msg.sender_did}: {msg.payload!r}")
            break

        # Bob leaves.
        await scp.context_leave(ctx._raw_handle, bob.did)

        # Alice closes the context.
        await scp.context_close(ctx._raw_handle, alice.did)


if __name__ == "__main__":
    asyncio.run(main())

File: bindings/typescript/examples/basic-messaging.ts
/**
 * Basic messaging: create identities, create a context, send a message, and
 * receive it on a subscriber callback.
 *
 * Post-Phase-4 (ADR-048): every SDK call routes through an explicit `SCP`
 * instance. Construct one at process start, invoke bridge operations as
 * `scp.<method>(...)`, and call `scp.shutdown(...)` to drain in-flight
 * work on exit.
 */

import { SCP } from "../src/index";

async function main(): Promise<void> {
  const scp = new SCP({ storage: { type: "in_memory" } });
  try {
    // Create two identities (in_memory custody for examples).
    const alice = await scp.identityCreate("in_memory");
    const bob = await scp.identityCreate("in_memory");
    console.log(`Alice DID: ${alice.did}`);
    console.log(`Bob DID: ${bob.did}`);

    // Alice creates a context and Bob joins it.
    const ctx = await scp.contextCreate(
      alice,
      JSON.stringify({
        ceiling: ["messages:read", "messages:write", "member:invite"],
        memoryScope: "ephemeral",
        governance: "single_admin",
        ttl: 3600,
      }),
    );
    console.log(`Context ID: ${ctx.contextId}`);

    await scp.contextJoin(ctx._rawHandle, bob.did, null);

    // Subscribe Bob to incoming messages BEFORE Alice sends so we don't
    // race the first payload.
    let received: { senderDid: string; payload: Uint8Array } | null = null;
    await scp.contextSubscribe(ctx._rawHandle, bob.did, (raw) => {
      const msg = raw as { senderDid?: string; payload?: number[] | Uint8Array };
      if (msg?.senderDid !== undefined && msg.payload !== undefined) {
        const bytes = msg.payload instanceof Uint8Array ? msg.payload : new Uint8Array(msg.payload);
        received = { senderDid: msg.senderDid, payload: bytes };
      }
    });

    // Alice sends a message.
    await scp.contextSend(
      ctx._rawHandle,
      alice.did,
      new TextEncoder().encode("Hello Bob, this is Alice"),
      null,
    );

    // Give the subscription a chance to flush.
    await new Promise((resolve) => setTimeout(resolve, 50));
    if (received !== null) {
      const { senderDid, payload } = received as { senderDid: string; payload: Uint8Array };
      console.log(`Bob received from ${senderDid}: ${new TextDecoder().decode(payload)}`);
    }

    // Leave and close.
    await scp.contextLeave(ctx._rawHandle, bob.did);
    await scp.contextClose(ctx._rawHandle, alice.did);
  } finally {
    // Graceful bridge shutdown — drains in-flight tasks within 5 seconds.
    await scp.shutdown(5);
  }
}

main().catch((error: unknown) => {
  console.error("Demo failed:", error);
  process.exit(1);
});
